README.md

```diff
@@ -0,0 +1,443 @@
+This single `Module` can then be included by apps that require that functionality.
+
+### Testing
+
+Testing for a component should use `fxtest` to create the component.
+This focuses testing on the API surface of the component against which other components will be built.
+Per-function unit tests are, of course, also great where appropriate!
+
+Here's an example testing a component with a mocked dependency on `other`:
+
+```go
+func TestMyComponent(t *testing.T) {
+    var comp Component
+    var other other.Component
+    app := fxtest.New(t,
+        Module,              // use the real version of this component
+        other.MockModule,    // use the mock version of other
+        fx.Populate(&comp),  // get the instance of this component
+        fx.Populate(&other), // get the (mock) instance of the other component
+    )
+
+    // start and, at completion of the test, stop the components
+    defer app.RequireStart().RequireStop()
+
+    // cast `other` to its mock interface to call mock-specific methods on it
+    other.(other.Mock).SetSomeValue(10)                      // Arrange
+    comp.DoTheThing()                                        // Act
+    require.Equal(t, 20, other.(other.Mock).GetSomeResult()) // Assert
+}
+```
+
+If the component has a mock implementation, it is a good idea to test that mock implementation as well.
+
+# Conventions
+
+## Package paths
+
+ * `cmd/<appname>/main.go` -- entrypoint for each app
+ * `pkg/<pkgnname>/...` utility types and functions, ["plain old data"](https://en.wikipedia.org/wiki/Passive_data_structure) types
+ * `pkg/util/<utilname>/...` utility packages that are not in the form of a component
+ * `comp/...` components
+ * `comp/<bundlename>/...` component bundles
+
+## Subscriptions
+
+A common mode of interaction between two components is for one to subscribe to notifications from the other.
+The `pkg/util/subscriptions` package provides generic support for this.
+
+Subscriptions are usually static for the lifetime of an Agent, and should be made during the setup phase, before components have started.
+The documentation for components supporting subscriptions will make this clear.
+
+## Actors
+
+Components often take the form of an [actor](https://en.wikipedia.org/wiki/Actor_model): a dedicated goroutine that processes events in a loop.
+This approach requires no concurrency controls, since all activity takes place in one goroutine.
+It is also easy to test: start a goroutine, send it some events, and assert on the output.
+
+The `pkg/util/actor` package supports components that use the actor structure, including connecting them to the `fx` life cycle.
+
+## IPC API Commands
+
+Several commands, such as `agent status` or `agent config`, call the running Agent's IPC API and format the result.
+Components implementing this pattern should generally have two similar methods, such as `GetStatus` and `GetStatusRemote`.
```

Comment:
Shouldn't it be two implementations of the same component with `GetStatus`?
With the current interface, it means all other components down the line will need to know if they're instantiated from CLI/distant process or from an Agent.

If we strictly speak about CLI, I'm also not sure the component should be different whether a CLI exists for it or not. It would be cleaner to simply inject `Remote` variant of components in the command implementation.

Reply:
I haven't given too much thought to remote "versions" of components, since that was very much a stretch goal in the RFC.  A generalized solution where we can easily substitute local and remote implementations may be useful later, for example to allow an agent to span multiple processes.  But for the simple case of supporting IPC for the CLI, I think that's too much: it's known at build time whether we're local or remote.

In fact, maybe those should be two different components?  Or even, the `cmd/agent/status` package should be responsible for unmarshaling and displaying the output?

Reply:
Yes I believe `cmd/agent/status` should be responsible for unmarshaling and displaying the output.
Maybe for CLI we know, but for remote tagger, for instance, we don't and we'll need it right when we'll make it a component.
I believed there's not much work on the framework side to handle remote or not, just a general recommendation that component interface **may** be implemented with remote calls in the future.

Reply:
Good point about unmarshaling in cmd/ -- I'll do that.